packages/service-library/src/servicelib/rabbitmq/_client.py

```diff
@@ -20,6 +21,62 @@
 
 _DEFAULT_PREFETCH_VALUE: Final[int] = 10
 _DEFAULT_RABBITMQ_EXECUTION_TIMEOUT_S: Final[int] = 5
+_HEADER_X_DEATH: Final[str] = "x-death"
+
+_DEFAULT_UNEXPECTED_ERROR_RETRY_DELAY_S: Final[float] = 1
+_DEFAULT_UNEXPECTED_ERROR_MAX_ATTEMPTS: Final[NonNegativeInt] = 15
+
+_DELAYED_EXCHANGE_NAME: Final[str] = "delayed_{exchange_name}"
+
+
+def _get_x_death_count(message: aio_pika.abc.AbstractIncomingMessage) -> int:
+    count: int = 0
+
+    x_death: list[dict[str, Any]] = message.headers.get(_HEADER_X_DEATH, [])
+    if x_death:
+        count = x_death[0]["count"]
+
+    return count
+
+
+async def _on_message(
+    message_handler: MessageHandler,
+    max_retries_upon_error: int,
+    message: aio_pika.abc.AbstractIncomingMessage,
+) -> None:
+    async with message.process(requeue=True, ignore_processed=True):
+        try:
+            with log_context(
+                _logger,
+                logging.DEBUG,
+                msg=f"Received message from {message.exchange=}, {message.routing_key=}",
+            ):
+                if not await message_handler(message.body):
+                    # NOTE: puts message to the Dead Letter Exchange
+                    await message.nack(requeue=False)
```

Comment:
hmm I think there is still something missing here.
If in my use case I don't raise wildly exceptions, this will retry forever (as we just discussed right?), even with a delay.

Sorry I thought we agreed on doing the exact same between returning False and raising an exception?
Here you do 2 different things.

My proposal would go like something like so:
```python
async def _safe_message_handler(message_handler, message_body) -> bool:
    try:
        return message_handler(message_body)
    except Exception:
        return False


async def _on_message(
    message_handler: MessageHandler,
    max_retries_upon_error: int,
    message: aio_pika.abc.AbstractIncomingMessage,
) -> None:
     async with message.process(requeue=True, ignore_processed=True):
        try:
            with log_context(
                _logger,
                logging.DEBUG,
                msg=f"Received message from {message.exchange=}, {message.routing_key=}",
            ):
                if not await _safe_message_handler(message.body):
                    count = _get_x_death_count(message)
                    _logger.warning(
                        (
                            "Retry [%s/%s] for handler '%s', which raised "
                            "an unexpected error caused by message_id='%s'"
                        ),
                        count,
                        max_retries_upon_error,
                        message_handler,
                        message.message_id,
                    )

                    if count < max_retries_upon_error:
                        # NOTE: puts message to the Dead Letter Exchange
                        await message.nack(requeue=False)
                    else:
                        _logger.exception(
                            "Handler '%s' is giving up on message '%s' with body '%s'",
                            message_handler,
                            message,
                            message.body,
                        )
                else:
                    await message.ack()

```

Reply:
Good point I've added a test and fixed it. Did it slightly differently. It should behave as expected now.